apps/backend/src/http-ping-bucket/aggregate/http-ping-bucket-aggregate.service.ts

```diff
@@ -0,0 +1,146 @@
+import { ClickHouseClient } from '@clickhouse/client';
+import { Injectable } from '@nestjs/common';
+import { addHours, subDays, subHours } from 'date-fns';
+import { ClickhouseUtils } from 'src/clickhouse/clickhouse.utils';
+import { HttpPingBucketEntity } from '../core/entities/http-ping-bucket.entity';
+import {
+  HttpPingBucketNormalized,
+  HttpPingBucketSerialized,
+} from '../core/entities/http-ping-bucket.interface';
+import { HttpPingBucketSerializer } from '../core/entities/http-ping-bucket.serializer';
+import { BucketGrouping, HttpPingBucketReadService } from '../read/http-ping-bucket-read.service';
+
+@Injectable()
+export class HttpPingBucketAggregateService {
+  constructor(
+    private readonly clickhouse: ClickHouseClient,
+    private readonly httpPingBucketReadService: HttpPingBucketReadService,
+  ) {}
+
+  public async getBucketsForMonitor(
+    monitorId: string,
+    period: '24h' | '4d' | '90d' = '24h',
+  ): Promise<(HttpPingBucketSerialized | null)[]> {
+    const periodConfig = this.getPeriodConfig(period);
+
+    const existingBuckets = await this.httpPingBucketReadService.readBucketsForMonitor(
+      monitorId,
+      periodConfig.fromDate,
+      periodConfig.grouping,
+    );
+
+    const completeBuckets = await this.createCompleteBuckets(
+      existingBuckets,
+      periodConfig.fromDate,
+      periodConfig.grouping,
+      periodConfig.expectedBucketCount,
+    );
+
+    return completeBuckets.map((bucket) =>
+      bucket ? HttpPingBucketSerializer.serialize(bucket) : null,
+    );
+  }
+
+  private getPeriodConfig(period: '24h' | '4d' | '90d') {
+    const subtractDaysUtc = (days: number) => {
+      return subDays(new Date(), days);
+    };
+
+    const configs = {
+      '24h': {
+        fromDate: addHours(subtractDaysUtc(1), 1),
+        grouping: BucketGrouping.Hour,
+        expectedBucketCount: 24,
+      },
+      '4d': {
+        fromDate: addHours(subtractDaysUtc(4), 1),
+        grouping: BucketGrouping.Hour,
+        expectedBucketCount: 96,
+      },
+      '90d': {
+        fromDate: subtractDaysUtc(89),
+        grouping: BucketGrouping.Day,
+        expectedBucketCount: 90,
+      },
+    };
+
+    return configs[period];
+  }
+
+  private async createCompleteBuckets(
+    existingBuckets: HttpPingBucketNormalized[],
+    fromDate: Date,
+    grouping: BucketGrouping,
+    expectedBucketCount: number,
+  ): Promise<(HttpPingBucketNormalized | null)[]> {
+    const completeBuckets = HttpPingBucketNormalized.fromExisting(
+      existingBuckets,
+      fromDate,
+      grouping,
+      expectedBucketCount,
+    );
+
+    if (completeBuckets[0] === null) {
+      completeBuckets[0] = await this.tryAddVirtualBucket(grouping);
+    }
+
+    return completeBuckets;
+  }
+
+  private async tryAddVirtualBucket(
```

Comment:
IMO `Add` suggests that we are adding this somewhere. At least I understood it this way. Seems like we are not adding it anywhere. We are just trying to construct it

Reply:
changed to `tryCreateVirtualBucket`